src/sage/schemes/elliptic_curves/homset.py

```diff
@@ -0,0 +1,234 @@
+        defined over the same base ring.
+
+        TESTS::
+
+            sage: E = EllipticCurve('37a1')                                             # needs sage.schemes
+            sage: Hom(E, E).__class__                                                   # needs sage.schemes
+            <class 'sage.schemes.elliptic_curves.homset.EllipticCurveHomset_with_category'>
+
+        ::
+
+            sage: E1 = EllipticCurve(j=42)
+            sage: E2 = EllipticCurve(j=43)
+            sage: Hom(E1, E2)
+            Additive group of elliptic-curve morphisms
+              From: Elliptic Curve defined by y^2 = x^3 + 5901*x + 1105454 over Rational Field
+              To:   Elliptic Curve defined by y^2 + x*y = x^3 + x^2 + 1510*x - 140675 over Rational Field
+            sage: Hom(E1, E2) in CommutativeAdditiveGroups()
+            True
+            sage: Hom(E1, E2) in Rings()
+            False
+            sage: End(E1) in CommutativeRings()
+            True
+
+        ::
+
+            sage: E0 = EllipticCurve(GF(419), [1,0])
+            sage: EE0 = E0.change_ring(GF(419^2))
+            sage: End(E0) in CommutativeRings()
+            True
+            sage: End(EE0) in CommutativeRings()
+            False
+            sage: End(EE0) in Rings()
+            True
+            sage: E1 = EllipticCurve(GF(419), [1,1])
+            sage: EE1 = E1.change_ring(GF(419^2))
+            sage: Hom(E0, E1) in CommutativeAdditiveGroups()
+            True
+            sage: Hom(EE0, EE1) in CommutativeAdditiveGroups()
+            True
+            sage: Hom(E0, E1) in Rings()
+            False
+            sage: Hom(EE0, EE1) in Rings()
+            False
+        """
+        if not isinstance(E1, EllipticCurve_generic):
+            raise ValueError('domain must be an elliptic curve')
+        if not isinstance(E2, EllipticCurve_generic):
+            raise ValueError('codomain must be an elliptic curve')
+        base = E1.base_ring()
+        if base != E2.base_ring():
+            raise ValueError('domain and codomain must have the same base ring')
+
+        super().__init__(E1, E2, category=category, base=base)
+
+        from sage.categories.commutative_additive_groups import CommutativeAdditiveGroups
+        self._refine_category_(CommutativeAdditiveGroups())
+
+        if self.is_endomorphism_set():
+            from sage.categories.rings import Rings
+            self._refine_category_(Rings())
+
+            assert E1 == E2
+
+            # The endomorphism ring of an elliptic curve over a finite field
+            # is non-commutative if and only if the curve is supersingular and
+            # the Frobenius endomorphism equals a scalar.
+            if (base.characteristic() == 0 or E1.is_ordinary() or
+                    (base.is_finite() and E1.frobenius() not in ZZ)):
+                from sage.categories.commutative_rings import CommutativeRings
+                self._refine_category_(CommutativeRings())
```

Comment:
Rather than refine the category afterwards, it would be better to modify the `category` (for example, by `category & Rings()`) variable before passing it off to `super().__init__()`.

Reply:
I believe `category` here is the category of the *objects* for which this homset is defined, but `._refine_category_()` changes the category of the homset itself. I don't see an obvious way to pass a category for the homset itself to the parent constructor.

Reply:
You're right. It seems like the proper way to do this would be to implement the appropriate `Homsets` category to `Schemes_over_base` (or perhaps a more refined category is needed?). For an example, see the `Modules` category in `category/modules.py`. Refining categories can be somewhat brittle IIRC, although within an `__init__` is safer I believe.

Reply:
The correct category for this would probably be abelian varieties. I'd personally leave that for the future (when we hopefully have more examples than just elliptic curves!) unless there are real problems with the current approach. Do you insist on doing it the proper way?

Reply:
For this I feel a bit more insistent since there have been many subtle issues that have come up with using `_refine_category_`. To be fair, most of those involve refining something like `Zmod(5)` once a primality test is done, where elements could already exist. That being said, there are already multiple proper mechanisms for setting the correct categories for morphisms before the objects are created.

Reply:
Alright, I gave it a try. I'm not sure of anything I'm doing here, though, since it's my first time dealing with categories in Sage...

Reply:
From first glance at what is pushed, it looks good and tests seem to be passing.